product.go

```diff
@@ -72,12 +78,99 @@ type ProductsResource struct {
 	Products []Product `json:"products"`
 }
 
+// Pagination of results
+type Pagination struct {
+	NextPageOptions		*ListOptions
+	PreviousPageOptions *ListOptions
+}
+
 // List products
 func (s *ProductServiceOp) List(options interface{}) ([]Product, error) {
+	products, _, err := s.ListWithPagination(options)
+	return products, err
+}
+
+// List products and return pagination to retrieve next/previous results
+func (s *ProductServiceOp) ListWithPagination(options interface{}) ([]Product, Pagination, error) {
 	path := fmt.Sprintf("%s/%s.json", globalApiPathPrefix, productsBasePath)
 	resource := new(ProductsResource)
-	err := s.client.Get(path, resource, options)
-	return resource.Products, err
+	headers := http.Header{}
+
+	err := s.client.CreateAndDo("GET", path, nil, options, resource, &headers)
+
+	// Extract pagination from header
+	linkHeader := ""
+	_, ok := headers["Link"]
+	if ok {
+		linkHeader = headers["Link"][0]
+	}
+	pagination, err := extractPagination(linkHeader)
+
+	return resource.Products, pagination, err
+}
+
+// Extract pagination from the Link header 
+// Details on the format are here:
+// https://help.shopify.com/en/api/guides/paginated-rest-results
+func extractPagination(linkHeader string) (Pagination, error) {
+	pagination := Pagination{}
+	
+	if linkHeader == "" {
+		return pagination, nil
+	}
+
+	linkRegex := regexp.MustCompile(`^ *<([^>]+)>; rel="(previous|next)" *$`)
```

Comment:
This should be moved out to be a global value(still private) so that you only compile it once instead of on every function call. Also avoids potentially panicing during the function due to a bad regex (ie panic at (startup/in a test) instead of randomly when this is called).